### ISSUE TITLE
SOAP integration: set root element namespace

### ISSUE THREAD

**Issue by alexkwolfe** [activeprospect/leadconduit-integration-custom#41]
Some SOAP services require the XML namespace to be set on the root element. As is, the integration might produce a request body like this:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope 
  xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" 
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
  xmlns:tns="http://blah.com/services/Soap">
  <soap:Body>
    <getRecipientByAddress>
      <address>foo@bar.baz</address>
    </getRecipientByAddress>
  </soap:Body>
</soap:Envelope>
```

But the service requires the namespace on the root element, like this:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope 
   xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
   xmlns:tns="http://blah.com/services/Soap">
  <soap:Body>
    <tns:getRecipientByAddress xmlns:tns="http://blah.com/services/Soap">
      <address>foo@bar.baz</address>
    </tns:getRecipientByAddress>
  </soap:Body>
</soap:Envelope>
```

This can be accomplished by adding support in the request variables for setting the `overrideRootElement` configuration of the [node-soap](https://github.com/vpulim/node-soap) library.

``` coffeescript
options =
    valueKey: '#value'
    forceSoap12Headers: vars.version?.trim() == '1.2'
    disableCache: true
    # This is not currently supported, but needs to be
    overrideRootElement:
      namespace: "tns"
      xmlnsAttributes: [
        { name: "xmlns:tns", value: "http://skylist.com/services/Soap" }
      ]
```

**Comment by mrjuicebox** [comment]
https://activeprospect.zendesk.com/agent/tickets/10860 is blocked by this.

What's an ETA I can provide to DMi?

**Comment by alexkwolfe** [comment]
We will discuss at our integrations meeting at 1:30 central today.

**Comment by alexkwolfe** [comment]
@jgrayson676, is this something you have time to look at?

**Comment by cgrayson** [comment]
I see that `options.xmlnsAttributes` is an array. Should there be support for multiple name/value pairs, @alexkwolfe ?

**Comment by alexkwolfe** [comment]
Yes. I suppose there should be support for multiple name value pairs.

**Comment by cgrayson** [comment]
Poking around with some local tests on the `fix/41` branch, I'm concerned that the behavior of those options isn't exactly as described, @alexkwolfe. 

In particular, using a `root_namespace_prefix` of 'tns', as in [this test](https://github.com/activeprospect/leadconduit-integration-custom/blob/a08999b1a8fec0e3fa70371b9096db5229ac1e3c/spec/soap-spec.coffee#L639-L660), yields this XML: 

```xml
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
               xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="http://donkey/ws.asmx/"
               xmlns:tm="http://microsoft.com/wsdl/mime/textMatching/">
  <soap:Body>
    <AddLead xmlns:tns="http://donkey/ws.asmx/"></AddLead>
  </soap:Body>
</soap:Envelope>
```

Looking at the `<AddLead xmlns:tns="http://donkey/ws.asmx/"></AddLead>` line, it _does_ have the added `xmlns:tns=` attribute, but it does _not_ prepend anything to the tag name. From your original writeup on this issue, I expected that should be `<tns:AddLead …`. 

Whether that's because the code on that branch isn't right, or if it's even necessary, I'm not sure. 

Also, I found that using another value besides 'tns', e.g., 'foo', _does_ make it insert the prefix. So maybe 'tns' is treated specially? I don't know.

```xml
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
               xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="http://donkey/ws.asmx/"
               xmlns:tm="http://microsoft.com/wsdl/mime/textMatching/">
  <soap:Body>
    <foo:AddLead xmlns:tns="http://donkey/ws.asmx/"></foo:AddLead>
  </soap:Body>
</soap:Envelope>
```

@andela-joladimeji - you might check with Jeremy or Paul, in reference to that ticket, to see if that `tns:` prefix is necessary for this customer.

**Comment by kellym** [comment]
Fixed with PR #50.
